client/network/src/persist_peers.rs

```diff
@@ -0,0 +1,252 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+	collections::{HashMap, HashSet},
+	fmt,
+	future::Future,
+	io,
+	path::{Path, PathBuf},
+	pin::Pin,
+	task::{Context, Poll},
+	time::{Duration, Instant},
+};
+
+use futures::{channel::oneshot, FutureExt};
+use lru::LruCache;
+
+use sc_peerset::PeersetHandle;
+
+use crate::{utils, Multiaddr, PeerId};
+
+type BoxedFuture<T> = Pin<Box<dyn Future<Output = T> + Send + 'static>>;
+type Never = std::convert::Infallible;
+type ProtocolType = String;
+
+const FLUSH_INTERVAL: Duration = Duration::from_secs(5);
+const PEER_ADDRS_CACHE_SIZE: usize = 100;
+
+const PEER_ADDRS_FILENAME: &str = "peer-addrs.json";
+const PEER_SETS_FILENAME: &str = "peer-sets.json";
```

Comment:
Why not join them?

Reply:
It was handy to hook to the address-resolution in `sc-network::discovery::DiscoveryBehaiour`, and to persist and restore peer-sets in `sc-network::protocol::Protocol`. The two things operate separately — they use two separate files.